ergotree-interpreter/src/sigma_protocol/prover.rs

```diff
@@ -1010,11 +1008,29 @@ fn step9_real_dh_tuple<P: Prover + ?Sized>(
                     pi
                 )));
             }
-            None => {
-                return Err(ProverError::NotYetImplemented(
-                    "when secret not found".to_string(),
-                ));
-            }
+            None => match hints_bag
+                .real_proofs()
+                .iter()
+                .find(|c| c.position == dhu.position)
+            {
+                Some(proof) => {
+                    let unchecked_tree = proof.clone().unchecked_tree;
+                    if let UncheckedTree::UncheckedLeaf(UncheckedLeaf::UncheckedDhTuple(
+                        unchecked_dht,
+                    )) = unchecked_tree
+                    {
+                        unchecked_dht.second_message
+                    } else {
+                        let z =
```

Comment:
According to Scala reference implementation in this case we should return an error. See https://github.com/ScorexFoundation/sigmastate-interpreter/blob/795c176debfe105646a7deef3f01bd537b79c81d/sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala#L535
`asInstanceOf` will just throw if casting is impossible.

Reply:
Yes Right, i think in the new commit the problem is solved.